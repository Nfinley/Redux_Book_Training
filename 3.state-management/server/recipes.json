[
  {
    "id": 10,
    "name": "Home made cookies",
    "author": {
      "id": 6123,
      "name": "The Cookie Monster"
    }
  },
  {
    "id": 11,
    "name": "Cold sangria",
    "author": {
      "id": 513,
      "name": "Elmo the Bird"
    }
  },
  {
    "id": 12,
    "name": "Lukewarm sangria",
    "author": {
      "id": 513,
      "name": "Elmo the Bird"
    }
  }
]