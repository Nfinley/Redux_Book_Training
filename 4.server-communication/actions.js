const fetchRecipes = () => ({
  type: API_REQUEST,
  payload: {
    url: 'db.json',
    next: FETCH_RECIPES
  }
});
