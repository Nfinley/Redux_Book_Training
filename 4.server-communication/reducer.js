const initialState = {
  loading: false,
  list: []
};

const recipesReducer = (state = initialState, action) => {
  switch (action.type) {

    case FETCH_RECIPES.PENDING:
      return Object.assign({}, state, {
        loading: true
      });

    case FETCH_RECIPES.SUCCESS:
      return Object.assign({}, state, {
        loading: false,
        list: action.payload.recipes.map((recipe) => recipe.name)
      });

    case FETCH_RECIPES.ERROR:
      return Object.assign({}, state, {
        loading: false
      });

    default:
      return state;
  }
};
