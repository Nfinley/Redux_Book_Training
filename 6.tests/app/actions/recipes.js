import 'whatwg-fetch';
import { createAction } from 'redux-actions';
import * as actions from 'consts';

export const addRecipe = title => ({
  type: actions.ADD_RECIPE,
  payload: title || "Default"
});

export const delRecipe = title => ({
  type: actions.DEL_RECIPE,
  payload: title
});

export const setRecipes = recipes => ({
  type: actions.SET_RECIPES,
  payload: recipes
});

export const fetchRecipes = () => ({
  type: actions.API,
  payload: {
    success: setRecipes,
    url: 'recipes.json'
  }
});

export const setRecipe = (id, data) => ({
  type: actions.SET_RECIPE,
  payload: {
    id,
    data
  }
});

export const fetchError = error => ({
  type: actions.FETCH_ERROR,
  error: true,
  payload: error
});

export const fetchRecipe = id => dispatch => {
  return fetch('recipe/' + id)
    .then(response => response.json())
    .then(json => dispatch(setRecipe(id, json)))
    .catch(({ status, statusText }) => dispatch(fetchError({ status, statusText })))
};