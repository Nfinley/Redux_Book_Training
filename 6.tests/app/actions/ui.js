import { createAction } from 'redux-actions';
import * as actions from 'consts';

export const apiStarted = createAction(actions.API_STARTED);
export const apiFinished = createAction(actions.API_FINISHED);
export const apiError = createAction(actions.API_ERROR);