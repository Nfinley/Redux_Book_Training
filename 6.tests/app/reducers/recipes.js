import { handleActions } from 'redux-actions';
import { ADD_RECIPE, DEL_RECIPE, SET_RECIPES } from 'consts.js';

const initialState = [];

export default handleActions({
  [ADD_RECIPE]:  (recipes, action) => recipes.concat({ title: action.payload }),
  [DEL_RECIPE]:  (recipes, action) => recipes.filter(recipe => recipe.title !== action.payload),
  [SET_RECIPES]: (recipes, action) => action.payload
}, initialState);