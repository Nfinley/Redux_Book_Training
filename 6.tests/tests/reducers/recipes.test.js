import reducer from 'reducers/recipes';
import deepFreeze from 'deep-freeze';
import { addRecipe, delRecipe } from 'actions/recipes';

const initialState = deepFreeze(reducer(undefined, { type: 'INIT' }));

describe('recipes reducer', () => {

  it('should handle unknown actions', () => {
    expect(reducer(initialState, { type: 'FAKE' })).toBe(initialState);
  });

  describe('add action', () => {
    it('should add recipe to empty list', () => {
      expect(reducer(initialState, addRecipe('test'))).toMatchSnapshot()
    });

    it('should add recipe to a non-empty list', () => {
      const nonEmptyState = deepFreeze(reducer(initialState, addRecipe('first')));

      expect(reducer(nonEmptyState, addRecipe('test'))).toMatchSnapshot();
    });
  });

  describe('delete action', () => {
    // Create list of 3 recipes
    const baseState = deepFreeze(
      ['first', 'second', 'third']
        .reduce((state, recipe) => reducer(state, addRecipe(recipe)), initialState)
    );

    it('should delete recipe when exists', () => {
      expect(reducer(baseState, delRecipe('second'))).toMatchSnapshot();
    });

    it('should not delete recipe when it doesn\'t exist', () => {
      expect(reducer(baseState, delRecipe('fourth'))).toMatchSnapshot();
    });
  });
});