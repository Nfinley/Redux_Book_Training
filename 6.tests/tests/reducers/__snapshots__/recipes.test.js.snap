exports[`recipes reducer add action should add recipe to a non-empty list 1`] = `
Array [
  Object {
    "title": "first",
  },
  Object {
    "title": "test",
  },
]
`;

exports[`recipes reducer add action should add recipe to empty list 1`] = `
Array [
  Object {
    "title": "test",
  },
]
`;

exports[`recipes reducer delete action should delete recipe when exists 1`] = `
Array [
  Object {
    "title": "first",
  },
  Object {
    "title": "third",
  },
]
`;

exports[`recipes reducer delete action should not delete recipe when it doesn't exist 1`] = `
Array [
  Object {
    "title": "first",
  },
  Object {
    "title": "second",
  },
  Object {
    "title": "third",
  },
]
`;
